spec/expression.dd

```diff
@@ -261,6 +261,18 @@ $(GNAME AssignExpression):
     occurs. The resulting expression is a modifiable lvalue.
     )
 
+	$(P Operands of an assign expression are evaluated before the assign expression. 
+	The expression)
+        --------------
+        a op= b
+        --------------
+
+        is exactly the same as:
```

Comment:
@thewilsonator I suppose this would be a more accurate wording:
```suggestion
        is rewritten to:
```
Which would address the question of side effects.

Reply:
Is it though? I thought it would be rewritten to `auto [ref] __tmp = a; __tmp = __tmp op b;`, the number of evaluation is important.

Reply:
Ah, I thought that was clear. Yeah, I guess it makes sense to specify this more accurately.

Reply:
One exception to that rewrite though is array concatenation.
```
a ~= b
// a = a ~ b;
```
Because evaluating `a` changes its underlying size, so it would be conceptually
```
auto tmp = b
extend(&a)[beg .. end] = tmp;
```

Reply:
Why does evaluating `a` change its size ? Isn't the size change part of the execution of `~=`, not the evaluation of operands ?

Reply:
Because in `a op= b`, operand `a` can only be evaluated once.  Separating the size change from the evaluation means you'll have to compute it twice.

Reply:
> Is it though?

Nope, this is definitely wrong, see e.g. https://github.com/dlang/dmd/blob/a9e5b0db723f8edaf80df28f93257edf06d3f17b/test/runnable/evalorder.d#L59-L61.

Reply:
I translated that test for eval order  to the `~=` operator:
```

auto mul11ret3(T)(ref T s)
{
    s ~= 11;
    return [3];
}

void main()
{
	static auto test3(int[] val) { (val ~= 7) ~= mul11ret3(val); return val; }
	assert(test3([2]) == [2,7,11,3]);
}
```

DMD fails with [2,11,7]
gdc 10.2 fails with core.exception.OutOfMemoryError.

Reply:
> I translated that test for eval order to the `~=` operator:
> gdc 10.2 fails with core.exception.OutOfMemoryError.

Excellent, that means garbage is passed as the destination slot in `append(&dst, mul11ret3(val))` instead of the address to `val`.